tooling/camel-karaf-feature-maven-plugin/src/main/java/org/apache/camel/karaf/feature/maven/EnsureWrapBundleVersionMojo.java

```diff
@@ -0,0 +1,319 @@
+        for (int i = versionStartIndex + 1; i < chars.length; i++) {
+            if ('$' == chars[i]) {
+                return i - 1;
+            }
+        }
+
+        return chars.length - 1;
+    }
+
+    String getVersion(String Location) {
+        return getVersion(Location, getVersionStartIndex(Location), getVersionEndIndex(Location));
+    }
+
+    String getVersion(String location, int versionStartIndex, int versionEndIndex) {
+        return location.substring(versionStartIndex, versionEndIndex + 1);
+    }
+
+    String getValidVersion(String location, String version) throws Exception {
+        if (version.charAt(0) == '$') {
+            throw new Exception("Maven version placeholder '%s' wasn't resolved".formatted(version));
+        }
+
+        try {
+            // Test if version will work in Karaf
+            new Version(version);
+        } catch (Exception e) {
+
+            // TODO: only use cleanVersion if the artifact is non-osgi
+            String cleanVersion = VersionCleaner.clean(version);
+            try {
+                // Test if clean version will work in Karaf again!
+                new Version(cleanVersion);
+
+                getLog().debug(
+                        "Bundle location '%s' will be set with Bundle-Version '%s', the output of org.apache.felix.utils.version.VersionCleaner.clean(%s)"
+                                .formatted(location, cleanVersion, version));
+                return cleanVersion;
+
+            } catch (Exception newException) {
+                throw new Exception("Version '%s' is not OSGi compliant".formatted(cleanVersion), newException);
+            }
+        }
+        return version;
+    }
+
+    String updateExistingVersion(String location, String bundleVersioHeader) throws Exception {
+        int versionHeaderStartIndex = location.indexOf(BUNDLE_VERSION);
+        int versionHeaderEndIndex = getBundleVersionHeaderEndIndex(location, versionHeaderStartIndex);
+
+        // BUNDLE_VERSION.length() + 1 will include '='
+        String currentVersion = location.substring(versionHeaderStartIndex + BUNDLE_VERSION.length() + 1,
+                versionHeaderEndIndex + 1);
+        if (currentVersion.charAt(0) == '$' || !currentVersion.equals(getValidVersion(location, currentVersion))) {
+            String currentBundleVersionHeader = location.substring(versionHeaderStartIndex, versionHeaderEndIndex + 1);
+
+            return location.replaceAll(Pattern.quote(currentBundleVersionHeader),
+                    Matcher.quoteReplacement(bundleVersioHeader));
+        }
+
+        return location;
+    }
+
+    /**
+     * @param location
+     * @param versionHeaderStartIndex
+     * @return wrap protocol Bundle-Version header last char index, inclusive
+     */
+    int getBundleVersionHeaderEndIndex(String location, int versionHeaderStartIndex) {
+        char[] chars = location.toCharArray();
```

Comment:
Warning `toCharArray` creates a new char array while here your goal is only to read some characters in the String, so use `charAt` instead

Reply:
Yes, I've forgotten about this 👍